active/0000-addon-instrumentation-experimental-hooks.md

```diff
@@ -0,0 +1,373 @@
+passed, nodes for which `until` returns `true` will not be yielded, nor will
+their subtrees.
+
+```js
+// for a tree
+//  TreeMerger
+//    |- Babel_1
+//    |- Babel_2
+//    |--|- Funnel
+for (n of node.preOrderIterator()) {
+  console.log(n.label.name);
+}
+// prints
+//
+// TreeMerger
+// Babel_1
+// Babel_2
+// Funnel
+
+
+for (n of node.preOrderIterator(x => x.label.name === 'Babel_2')) {
+  console.log(n.label.name);
+}
+// prints
+//
+//  TreeMerger
+//  |- Babel_1
+```
+
+##### postOrderIterator(until)
+
+Returns an iterator that yields every node in the subtree rooted at this node.
+Nodes are yielded after their children.  If the optional parameter `until` is
+passed, nodes for which `until` returns `true` will not be yielded, nor will
+their subtrees.
+
+Example:
+```js
+// for a tree
+//  TreeMerger
+//    |- Babel_1
+//    |- Babel_2
+//    |--|- Funnel
+for (n of node.postOrderIterator()) {
+  console.log(n.label.name);
+}
+// prints
+//
+// Funnel
+// Babel_2
+// Babel_1
+// TreeMerger
+```
+
+##### ancestorsIterator()
+
+Returns an iterator that yields every ancestor of this node, in order from the
+node's parent to the root.  The node itself is not yielded.
+
+Example:
+```js
+// for a tree
+//  TreeMerger
+//    |- Babel_1
+//    |- Babel_2
+//    |--|- Funnel
+//
+// on node Funnel
+for (n of node.ancestorsIterator()) {
+  console.log(n.label.name);
+}
+// prints
+//
+// Babel_2
+// TreeMerger
+```
+
+##### statsIterator()
+
+Returns an iterator that yields `[name, value]` pairs of stat names and values.
+
+Example:
+```js
+  //  for a typical broccoli node 
+  for ([statName, statValue] of node.statsIterator()) {
+    console.log(statName, statValue);
+  }
+  // prints
+  //
+  // "time.self" 64232794
+  // "fs.statSync.count" 40
+  // "fs.statSync.time" 401232123
+  // ...
+  ```
+
+### Load Instrumentation
+
+#### Hook
+
+An addon that implements `loadInstrumentation` will have this hook invoked when
+instrumentation is enabled.
+
+```js
+module.exports = {
+  name: 'my-great-addon',
+
+  loadInstrumentation(loadInstrumentation) {
+    // format of loadInstrumentation outlined below
+  }
+};
+```
+
+#### loadInstrumentation format
+
+`loadInstrumentation` is an object with two properties, `summary` and
+`loadTree`.
+
+##### summary
+
+The summary object has the following shape:
+
+```js
+{
+  totalTime,
+  command,
+  platform: {
+    name,
+  },
+}
+```
+
+The properties are as follows:
+
+- `summary.totalTime` The total time from when we started measuring (as close to
+  node process startup as possible) until the start of the build, for `build`
+  and `serve` commands, or the end of the process, for other commands.
+- `summary.command` The name of the command that was run.
+- `summary.platform.name` The value of `process.platform`
+
+##### loadTree
+
+The `loadTree` API is exactly the same as the `buildTree` API.
+
+
+## Additional Comments
+
+# How We Teach This
```

Comment:
I know this is something not used by all addon authors but it would be nice to have an example of how to use the API in a real world scenario in the API docs and on ember-cli website. It helps users of `ember-cli` understand the use of this public API better.